librz/core/canalysis.c

```diff
@@ -6688,6 +6688,90 @@ RZ_IPI char *rz_core_analysis_function_signature(RzCore *core, RzOutputMode mode
 	return signature;
 }
 
+static RzAnalysisBlock *find_block_at_xref_addr(RzCore *core, ut64 addr) {
+	RzList *blocks = rz_analysis_get_blocks_in(core->analysis, addr);
+	if (!blocks) {
+		return NULL;
+	}
+	RzAnalysisBlock *block = NULL;
+	RzListIter *bit;
+	RzAnalysisBlock *block_cur;
+	rz_list_foreach (blocks, bit, block_cur) {
+		if (rz_analysis_block_op_starts_at(block_cur, addr)) {
+			block = block_cur;
+			break;
+		}
+	}
+	if (block) {
+		rz_analysis_block_ref(block);
+	}
+	rz_list_free(blocks);
+	return block;
+}
+
+static void relocation_function_process_noreturn(RzCore *core, RzAnalysisFunction *fcn, ut64 opsize, ut64 reladdr, ut64 addr) {
+	rz_analysis_noreturn_add(core->analysis, NULL, reladdr);
+	// Chop the function
+	rz_analysis_function_chop_at(fcn, addr + opsize);
+	if (analyze_noreturn_function(core, fcn)) {
+		fcn->is_noreturn = true;
+		rz_analysis_noreturn_add(core->analysis, NULL, fcn->addr);
+	}
+}
+
+static void relocation_noreturn_process(RzCore *core, RzList *noretl, RzAnalysisFunction *fcn, RzBinReloc *rel, ut64 opsize, ut64 addr) {
+	RzListIter *iter3;
+	char *noret;
+	if (rel->import) {
+		rz_list_foreach (noretl, iter3, noret) {
+			if (!strcmp(rel->import->name, noret)) {
+				relocation_function_process_noreturn(core, fcn, opsize, rel->vaddr, addr);
+			}
+		}
+	} else if (rel->symbol) {
+		rz_list_foreach (noretl, iter3, noret) {
+			if (!strcmp(rel->symbol->name, noret)) {
+				relocation_function_process_noreturn(core, fcn, opsize, rel->symbol->vaddr, addr);
+			}
+		}
+	}
+}
+
+#define CALL_BUF_SIZE 32
+
+RZ_API void rz_core_analysis_propagate_noreturn_relocs(RzCore *core, ut64 addr) {
+	// find known noreturn functions to propagate
+	RzListIter *iter;
+	RzAnalysisFunction *fcn;
+	RzList *noretl = rz_types_function_noreturn(core->analysis->sdb_types);
+	rz_list_foreach (core->analysis->fcns, iter, fcn) {
+		RzListIter *iter2;
+		RzList *refs = rz_analysis_function_get_refs(fcn);
+		RzAnalysisRef *ref;
+		rz_list_foreach (refs, iter2, ref) {
```

Comment:
I think this kind of iteration scheme is not what you want, since actually from the caller side you don't care about functions, but just about basic blocks. In particular `rz_analysis_function_get_refs()` should not be used for anything but user output and callgraphs I think.

So what I would do instead is to either:
* iterate all relocs { check if there is an xref from there, get the basic block containing the instruction, chop }
  I guess this won't work out because you can't easily get the instruction start address from the reloc as it can start in the middle of the instruction.
* iterate all matching xrefs globally { get the basic block at the "from" addr, get the op+reloc at the from addr, proceed as currently }
  I think this will work best.